### PR TITLE
bpo-24821: Fixed the slowing down to 25 times in the searching of some

### PR DESCRIPTION
**Pull request by serhiy-storchaka**
unlucky Unicode characters.

### CODE REVIEW

File: Objects/stringlib/fastsearch.h [hunk 1]
```diff
@@ -48,25 +54,35 @@ STRINGLIB(find_char)(const STRINGLIB_CHAR* s, Py_ssize_t n, STRINGLIB_CHAR ch)
...
            false positives */
+        const STRINGLIB_CHAR *s1, *e1;
         unsigned char needle = ch & 0xff;
         /* If looking for a multiple of 256, we'd have too
            many false positives looking for the '\0' byte in UCS2
            and UCS4 representations. */
-        if (needle != 0) {
-            while (p < e) {
+        if (needle != 0)
+            while (e - p > MEMCHR_CUT_OFF) {
```
Comment (zhangyangyu) [inline Objects/stringlib/fastsearch.h]: Maybe `do ... while ...` is better here? The first time `e - p > MEMCHR_CUT_OFF` is always true.

File: Objects/stringlib/fastsearch.h [hunk 2]
```diff
@@ -86,34 +102,46 @@ STRINGLIB(rfind_char)(const STRINGLIB_CHAR* s, Py_ssize_t n, STRINGLIB_CHAR ch)
        it doesn't seem as optimized as memchr(), but is still quite
        faster than our hand-written loop below */
 
-    if (n > 10) {
+    if (n > MEMCHR_CUT_OFF) {
 #if STRINGLIB_SIZEOF_CHAR == 1
-        p = memrchr(s, ch, n);
+        p = memrchr(s, ch, n * sizeof(STRINGLIB_CHAR));
```
Comment (zhangyangyu) [inline Objects/stringlib/fastsearch.h]: Why this?

File: Objects/stringlib/fastsearch.h [hunk 3]
```diff
@@ -48,25 +54,35 @@ STRINGLIB(find_char)(const STRINGLIB_CHAR* s, Py_ssize_t n, STRINGLIB_CHAR ch)
...
+                if (e - p <= MEMCHR_CUT_OFF)
+                    break;
+                e1 = p + MEMCHR_CUT_OFF;
+                while (p != e1) {
+                    if (*p == ch)
+                        return (p - s);
+                    p++;
+                }
             }
-            return -1;
```
Comment (zhangyangyu) [inline Objects/stringlib/fastsearch.h:68]: How about `assert(e - p <= MEMCHR_CUT_OFF)` here?

File: Objects/stringlib/fastsearch.h (hunk 2 above)
Comment (serhiy-storchaka) [reply Objects/stringlib/fastsearch.h]: Ah, no good reason. Perhaps I tried to unify this with  the code for STRINGLIB_SIZEOF_CHAR != 1.

File: Objects/stringlib/fastsearch.h (hunk 1 above)
Comment (serhiy-storchaka) [reply Objects/stringlib/fastsearch.h]: Done.

File: Objects/stringlib/fastsearch.h (hunk 3 above)
Comment (serhiy-storchaka) [reply Objects/stringlib/fastsearch.h:68]: It is obvious if use a do-while loop.

Comment (zhangyangyu) [reply Objects/stringlib/fastsearch.h:68]: Ohh, yes. I comment with the previous code.

**Review by zhangyangyu — APPROVED** [review]

**Pull request merged**